File: solutions/494/main.py
from typing import List 

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        total = sum(nums)
        if abs(target) > total:
            return 0
        if (total + target) % 2 == 1:
            return 0
        sum_target = (total + target) // 2
        dp = [[0] * len(nums) for _ in range(sum_target+1)]

        # initialize
        for i in range(len(nums)):
            dp[0][i] = 1
        for i in range(sum_target+1):
            if i == nums[0]:
                dp[i][0] += 1

        for i in range(sum_target+1):
            for j in range(1, len(nums)):
                if i >= nums[j]:
                    dp[i][j] = dp[i][j-1] + dp[i-nums[j]][j-1]
                else:
                    dp[i][j] = dp[i][j-1]

        result = dp[sum_target][len(nums)-1]

        return result

if __name__ == "__main__": 
    nums = [1,1,1,1,1]
    target = 3
    s = Solution()
    assert s.findTargetSumWays(nums, target) == 5

File: solutions/494/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sun Nov 14 21:57:39 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <numeric> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        // sanity check
        int sum = 0;
        sum = accumulate(nums.begin(), nums.end(), sum);
        if (abs(target) > sum) return 0;
        if ((sum + target) % 2 == 1) return 0;

        // dp table
        int sumTarget = (sum + target) / 2;
        vector<vector<int>> dp(sumTarget+1, vector<int>(nums.size(), 0));

        // initialize
        for (int i = 0; i < nums.size(); ++i) {
            dp[0][i] = 1;
        }
        for (int i = 0; i <= sumTarget; ++i) {
            if (i == nums[0])
                dp[i][0] += 1;
        }

        // build dp table
        for (int i = 0; i <= sumTarget; ++i) {
            for (int j = 1; j < nums.size(); ++j) {
                if (i >= nums[j])
                    dp[i][j] = dp[i][j-1] + dp[i-nums[j]][j-1];
                else
                    dp[i][j] = dp[i][j-1];
            }
        }

        return dp[sumTarget][nums.size()-1];
    }
};

TEST(Test494, SimpleTest) {
    vector<int> nums{
        1,1,1,1,1
    };
    const int target = 3; 
    Solution s; 
    EXPECT_EQ(s.findTargetSumWays(nums, target), 5);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
